lib/rucio/api/replica.py

```diff
@@ -85,36 +86,54 @@ def declare_bad_file_replicas(replicas, reason, issuer, vo='def', force=False, *
-
-    issuer = InternalAccount(issuer, vo=vo)
-
     # make sure all elements are either strings or dicts, without mixing
-    type_ = type(replicas[0])
-    if any(not isinstance(r, type_) for r in replicas):
+
+    as_pfns = isinstance(replicas[0], str)
+    if any(isinstance(r, str) != as_pfns for r in replicas):
         raise exception.InvalidType('The replicas must be specified either as a list of PFNs (strings) or list of dicts')
 
+    rse_map = {}            # name -> id
+    rse_id_to_name = {}     # id -> name
+
+    issuer = InternalAccount(issuer, vo=vo)
+
     replicas_lst = replicas
-    if type_ is dict:
+    rse_ids_to_check = set()
+    if as_pfns:
+        scheme, rses_for_replicas, unknowns = get_pfn_to_rse(replicas, vo=vo)
+        if unknowns:
+            raise exception.ReplicaNotFound("Not all replicas found")
+        rse_ids_to_check = set(rses_for_replicas.keys())
+    else:
+        # repliac are given as dicts {name, scope, rse}
         replicas_lst = []           # need to create new list to convert replica["scope"] from strings to InternalScope objects
         for r in replicas:
             if "name" not in r or "scope" not in r or ("rse" not in r and "rse_id" not in r):
                 raise exception.InvalidType('The replica dictionary must include scope and either rse (name) or rse_id')
-            scope = InternalScope(r['scope'], vo=vo)
-            rse_id = r.get("rse_id") or rse_map.get(r['rse'])
-            if rse_id is None:
-                rse = r["rse"]
-                rse_map[rse] = rse_id = get_rse_id(rse=rse, vo=vo, session=session)
+            rse_id = r.get("rse_id")
+            rse = r.get("rse")
+            rse_id = rse_id or rse_map.get(rse) or get_rse_id(rse=rse, vo=vo, session=session)
+            rse = rse or rse_id_to_name.get(rse_id) or get_rse_name(rse_id, session=session)
+            rse_map[rse] = rse_id
+            rse_id_to_name[rse_id] = rse
             replicas_lst.append({
-                "scope": scope,
+                "scope": InternalScope(r['scope'], vo=vo),
                 "rse_id": rse_id,
                 "name": r["name"]
             })
+            rse_ids_to_check.add(rse_id)
 
-    rse_id_to_name = invert_dict(rse_map)   # RSE id -> RSE name
+    for rse_id in rse_ids_to_check:
+        try:
+            permitted = permission.has_permission(issuer=issuer, vo=vo, action='declare_bad_file_replicas',
+                                                  kwargs={"rse_id": rse_id},
+                                                  session=session)
+        except:
+            permitted = True
+
+        if not permitted:
+            raise exception.AccessDenied('Account %s can not declare bad replicas in RSE %s'
+                                         % (issuer, rse_id_to_name.get(rse_id, rse_id)))
 
     undeclared = replica.declare_bad_file_replicas(replicas_lst, reason=reason, issuer=issuer, status=BadFilesStatus.BAD,
```

Comment:
Hi @ivmfnal 

Now, that the replica_lst is always a dict with lfns and rse_id, we may get rid of dead code that uses lfns, and often does not work for other experiments, e.g. _declare_bad_file_replicas for cms 
https://github.com/rucio/rucio/blob/6eac5dcc30cb6dac427cbc5070d46c170ae7fedf/lib/rucio/core/replica.py#L265-L288

Reply:
this code may be "dead" for CMS, but maybe it is still used by other users.

Reply:
The replica_lst is not always a list of dictionaries. If the original list was a list of PFNs, then the list is not modified and passed to the `core.decalre_bad_file_replicas()` as is.